### PR TITLE
Jbuilder

### PR DESCRIPTION
**Pull request by hhugo**
This is a work in progress, resurrection of https://github.com/ocsigen/js_of_ocaml/pull/565,
making use of jbuilder, ocaml-migrate-parsetree, ppx_tools_versionned.

Js_of_ocaml is now split as follow
- js_of_ocaml-compiler: compiler & runtime
- js_of_ocaml-camlp4: camlp4 syntaxes (pa_js, pa_deriving)
- js_of_ocaml-ppx: ppx syntaxes (ppx_js, ppx_deriving_json)
- js_of_ocaml-ocamlbuild: ocamlbuild plugin
- js_of_ocaml-toplevel: jsoo_mkcmis, jsoo_mktop, js_of_ocaml.toplevel
- js_of_ocaml: js_of_ocaml lib with no dep on lwt, tyxml, ...
- js_of_ocaml-lwt: lwt support (Lwt_js, Lwt_events_js, ..), graphics, lwt-logger 
- js_of_ocaml-tyxml: support for tyxml

Async and ppx_driver support have disappeared and are/will be maintained by Janestreet (cc @diml) 

TODO
- [x] compile the toplevel example
- [x] compile other examples
- [x] generate the doc
- [ ] run tests
- [ ] fix META with compatibility layer (how much compat do we want to provide ?)

WAITING
- [x] release of jbuilder  > 1.0+beta2
- [x] release of ppx_tools_verioned > 5.0alpha
https://github.com/hhugo/ppx_tools_versioned.git

NEXT
- [ ] split a core js_of_ocaml lib that doesn't depend on a browser environment  
- [ ] make the compiler optionally depend on yojson with sourcemap support (https://github.com/ocsigen/js_of_ocaml/pull/568)

Should address the following issues
fix https://github.com/ocsigen/js_of_ocaml/issues/265
fix https://github.com/ocsigen/js_of_ocaml/issues/360
fix https://github.com/ocsigen/js_of_ocaml/issues/566
fix https://github.com/ocsigen/js_of_ocaml/issues/545
fix https://github.com/ocsigen/js_of_ocaml/issues/543
fix https://github.com/ocsigen/js_of_ocaml/issues/523

### CODE REVIEW

**Comment by Drup** [comment]
IIUC, this is a reboot of https://github.com/ocsigen/js_of_ocaml/pull/541 using jbuilder ? Interesting. Can you give a quick rundown of what us non-janstreetians need to know after this transition (from a practical point of view) ?

**Comment by hhugo** [comment]
The janestreetness does not really matter here ;)

You will be able to start using jbuilder with
```
opam install jbuilder
jbuilder external-lib-deps --missing @install
# install missing deps with opam
jbuilder build @install
```

It builds fast and support incremental build. which works well for development  
```
time jbuilder build @install -j 16
real	0m8.900s
user	0m41.739s
sys	0m7.069s

time jbuilder build @install -j 16
real	0m0.287s
user	0m0.089s
sys	0m0.062s
```
Using jbuilder makes managing multiple opam package for js_of_ocaml very easy.
```
jbuilder build @install --only js_of_ocaml-compiler
jbuilder install js_of_ocaml-compiler
```

Still need more work to fully replace the current build system.
I'll see if I can combine the two to support
- ocamldoc
- examples
- toplevel

It is also super convenient to tests multiple ocaml version / opam switch at once.
The following will build js_of_ocaml for both ocaml 4.03 and ocaml 4.04
```
cat jbuild-workspace.dev 
;; Install the following opam switches, copy this file as
;; jbuild-workspace and run:
;;
;; $ jbuilder build @install
;;
;; This will build js_of_ocaml against all these version of OCaml

;;(context ((switch 4.02.3)))
;;(context ((switch 4.03.0+32bit))
(context ((switch 4.03.0)))
(context ((switch 4.04.0)))
;;(context ((switch 4.05.0+beta2)))
;;(context ((switch 4.06.0+trunk)))
jbuilder build @install --workspace jbuild-workspace.dev
```

Note this PR also adds a dependency on ocaml-migrate-parsetree and ppx_tools_versioned.

**Comment by Drup** [comment]
That's the summary I was hoping for. Thanks! This looks really nice.

**Comment by rgrinberg** [comment]
yay @hhugo :100: :medal_military:

**Comment by Drup** [comment]
@hhugo I don't understand how ppx tests are run, if they are run at all.

**Comment by hhugo** [comment]
Tests have not been fully ported yet.

File: META.js_of_ocaml.template [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+package "tyxml" (
+  requires = "js_of_ocaml-tyxml"
+)
+
+package "graphics" (
+  requires = "js_of_ocaml-ppx.graphics"
+)
+
+package "log" (
+  requires = "js_of_ocaml-ppx.log"
```
Comment (vasilisp) [inline META.js_of_ocaml.template]: `js_of_ocaml-ppx.log` doesn't exist. Require `js_of_ocaml-lwt.logger`, or remove `js_of_ocaml.log` altogether?

Comment (hhugo) [inline META.js_of_ocaml.template]: I believe this is a typo. It should be js_of_ocaml-lwt.log

Comment (hhugo) [inline META.js_of_ocaml.template]: pushed a fix

**Pull request merged**